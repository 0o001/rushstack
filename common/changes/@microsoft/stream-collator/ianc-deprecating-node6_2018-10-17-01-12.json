{
  "changes": [
    {
      "packageName": "@microsoft/stream-collator",
      "comment": "Removing unused exports that don't work in Node 10.",
      "type": "major"
    }
  ],
  "packageName": "@microsoft/stream-collator",
  "email": "iclanton@users.noreply.github.com"
}