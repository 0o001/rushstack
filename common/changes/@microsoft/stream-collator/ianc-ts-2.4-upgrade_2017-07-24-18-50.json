{
  "changes": [
    {
      "packageName": "@microsoft/stream-collator",
      "comment": "Upgrade to TypeScript 2.4",
      "type": "patch"
    }
  ],
  "packageName": "@microsoft/stream-collator",
  "email": "iclanton@users.noreply.github.com"
}