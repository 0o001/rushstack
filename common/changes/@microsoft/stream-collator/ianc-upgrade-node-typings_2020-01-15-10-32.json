{
  "changes": [
    {
      "comment": "Upgrade Node typings to Node 10",
      "type": "minor",
      "packageName": "@microsoft/stream-collator"
    }
  ],
  "packageName": "@microsoft/stream-collator",
  "email": "iclanton@users.noreply.github.com"
}