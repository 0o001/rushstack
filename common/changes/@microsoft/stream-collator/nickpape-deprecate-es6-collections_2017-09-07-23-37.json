{
  "changes": [
    {
      "packageName": "@microsoft/stream-collator",
      "comment": "Deprecate @types/es6-coll ections in favor of built-in typescript typings 'es2015.collection' a nd 'es2015.iterable'",
      "type": "patch"
    }
  ],
  "packageName": "@microsoft/stream-collator",
  "email": "nickpape@users.noreply.github.com"
}