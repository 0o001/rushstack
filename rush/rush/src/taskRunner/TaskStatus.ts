/**
 * @file TaskStatus.ts
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 *
 * Enumeration defining potential states of a task: not started, executing, or completed
 */

enum TaskStatus {
  Ready = 1,
  Executing = 2,
  Success = 3,
  Failure = 4,
  Blocked = 5
}
export default TaskStatus;
