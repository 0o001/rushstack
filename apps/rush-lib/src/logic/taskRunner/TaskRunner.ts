// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as os from 'os';
import * as colors from 'colors';
import {
  StreamCollator,
  CollatedTerminal,
  StdioSummarizer,
  TerminalWritable,
  StdioWritable,
  TerminalChunkKind,
  CharMatcherTransform,
  CollatedWriter
} from '@rushstack/stream-collator';
import { AlreadyReportedError, NewlineKind } from '@rushstack/node-core-library';

import { Stopwatch } from '../../utilities/Stopwatch';
import { Task } from './Task';
import { TaskStatus } from './TaskStatus';
import { IBuilderContext } from './BaseBuilder';

export interface ITaskRunnerOptions {
  quietMode: boolean;
  parallelism: string | undefined;
  changedProjectsOnly: boolean;
  allowWarningsInSuccessfulBuild: boolean;
  destination?: TerminalWritable;
}

/**
 * A class which manages the execution of a set of tasks with interdependencies.
 * Initially, and at the end of each task execution, all unblocked tasks
 * are added to a ready queue which is then executed. This is done continually until all
 * tasks are complete, or prematurely fails if any of the tasks fail.
 */
export class TaskRunner {
  private _tasks: Task[];
  private _changedProjectsOnly: boolean;
  private _allowWarningsInSuccessfulBuild: boolean;
  private _buildQueue: Task[];
  private _quietMode: boolean;
  private _hasAnyFailures: boolean;
  private _hasAnyWarnings: boolean;
  private _parallelism: number;
  private _currentActiveTasks: number;
  private _totalTasks: number;
  private _completedTasks: number;

  private readonly _outputWritable: TerminalWritable;
  private readonly _colorsNewlinesTransform: CharMatcherTransform;
  private readonly _streamCollator: StreamCollator;

  private _terminal: CollatedTerminal;

  public constructor(orderedTasks: Task[], options: ITaskRunnerOptions) {
    const { quietMode, parallelism, changedProjectsOnly, allowWarningsInSuccessfulBuild } = options;
    this._tasks = orderedTasks;
    this._buildQueue = orderedTasks.slice(0);
    this._quietMode = quietMode;
    this._hasAnyFailures = false;
    this._hasAnyWarnings = false;
    this._changedProjectsOnly = changedProjectsOnly;
    this._allowWarningsInSuccessfulBuild = allowWarningsInSuccessfulBuild;

    // TERMINAL PIPELINE:
    //
    // streamCollator --> colorsNewlinesTransform --> StdioWritable
    //
    this._outputWritable = options.destination ? options.destination : StdioWritable.instance;
    this._colorsNewlinesTransform = new CharMatcherTransform({
      destination: this._outputWritable,
      normalizeNewlines: NewlineKind.OsDefault,
      removeColors: !colors.enabled
    });
    this._streamCollator = new StreamCollator({
      destination: this._colorsNewlinesTransform,
      onWriterActive: this._streamCollator_onWriterActive
    });
    this._terminal = this._streamCollator.terminal;

    const numberOfCores: number = os.cpus().length;

    if (parallelism) {
      if (parallelism === 'max') {
        this._parallelism = numberOfCores;
      } else {
        const parallelismInt: number = parseInt(parallelism, 10);

        if (isNaN(parallelismInt)) {
          throw new Error(`Invalid parallelism value of '${parallelism}', expected a number or 'max'`);
        }

        this._parallelism = parallelismInt;
      }
    } else {
      // If an explicit parallelism number wasn't provided, then choose a sensible
      // default.
      if (os.platform() === 'win32') {
        // On desktop Windows, some people have complained that their system becomes
        // sluggish if Rush is using all the CPU cores.  Leave one thread for
        // other operations. For CI environments, you can use the "max" argument to use all available cores.
        this._parallelism = Math.max(numberOfCores - 1, 1);
      } else {
        // Unix-like operating systems have more balanced scheduling, so default
        // to the number of CPU cores
        this._parallelism = numberOfCores;
      }
    }
  }

  private _streamCollator_onWriterActive = (writer: CollatedWriter | undefined): void => {
    if (writer) {
      this._completedTasks++;

      // Format a header like this, for a shell window with classic 80 columns
      //
      // ==[ @rushstack/the-long-thing ]=================[ 1 of 1000 ]==
      const headerWidth: number = 79;

      // leftPart: "==[ @rushstack/the-long-thing "
      const leftPart: string = colors.gray('==[') + ' ' + colors.cyan(writer.taskName) + ' ';
      const leftPartLength: number = 4 + writer.taskName.length + 1;

      // rightPart: " 1 of 1000 ]=="
      const completedOfTotal: string = `${this._completedTasks} of ${this._totalTasks}`;
      const rightPart: string = ' ' + colors.white(completedOfTotal) + ' ' + colors.gray(']==');
      const rightPartLength: number = 1 + completedOfTotal.length + 4;

      // middlePart: "]=================["
      const twoBracketsLength: number = 2;
      const middlePartLengthMinusTwoBrackets: number = Math.max(
        headerWidth - (leftPartLength + rightPartLength + twoBracketsLength),
        0
      );

      const middlePart: string = colors.gray(']' + '='.repeat(middlePartLengthMinusTwoBrackets) + '[');

      this._terminal.writeStdoutLine('\n' + leftPart + middlePart + rightPart + '\n');
    }
  };

  /**
   * Executes all tasks which have been registered, returning a promise which is resolved when all the
   * tasks are completed successfully, or rejects when any task fails.
   */
  public async executeAsync(): Promise<void> {
    this._currentActiveTasks = 0;
    this._completedTasks = 0;
    this._totalTasks = this._tasks.length;

    if (!this._quietMode) {
      const plural: string = this._tasks.length === 1 ? '' : 's';
      this._terminal.writeStdoutLine(`Selected ${this._tasks.length} project${plural}:`);
      this._terminal.writeStdoutLine(
        this._tasks
          .map((x) => `  ${x.name}`)
          .sort()
          .join('\n')
      );
      this._terminal.writeStdoutLine('');
    }

    this._terminal.writeStdoutLine(`Executing a maximum of ${this._parallelism} simultaneous processes...`);

    await this._startAvailableTasksAsync();

    this._printTaskStatus();

    if (this._hasAnyFailures) {
      throw new Error('Project(s) failed');
    } else if (this._hasAnyWarnings && !this._allowWarningsInSuccessfulBuild) {
      this._terminal.writeStderrLine('Project(s) succeeded with warnings');
      throw new AlreadyReportedError();
    }
  }

  /**
   * Pulls the next task with no dependencies off the build queue
   * Removes any non-ready tasks from the build queue (this should only be blocked tasks)
   */
  private _getNextTask(): Task | undefined {
    for (let i: number = 0; i < this._buildQueue.length; i++) {
      const task: Task = this._buildQueue[i];

      if (task.status !== TaskStatus.Ready) {
        // It shouldn't be on the queue, remove it
        this._buildQueue.splice(i, 1);
        // Decrement since we modified the array
        i--;
      } else if (task.dependencies.size === 0 && task.status === TaskStatus.Ready) {
        // this is a task which is ready to go. remove it and return it
        return this._buildQueue.splice(i, 1)[0];
      }
      // Otherwise task is still waiting
    }
    return undefined; // There are no tasks ready to go at this time
  }

  /**
   * Helper function which finds any tasks which are available to run and begins executing them.
   * It calls the complete callback when all tasks are completed, or rejects if any task fails.
   */
  private _startAvailableTasksAsync(): Promise<void> {
    const taskPromises: Promise<void>[] = [];
    let ctask: Task | undefined;
    while (this._currentActiveTasks < this._parallelism && (ctask = this._getNextTask())) {
      this._currentActiveTasks++;
      const task: Task = ctask;
      task.status = TaskStatus.Executing;

      task.stopwatch = Stopwatch.start();
      task.collatedWriter = this._streamCollator.registerTask(task.name);
      task.stdioSummarizer = new StdioSummarizer();

      taskPromises.push(this._executeTaskAndChainAsync(task));
    }

    return Promise.all(taskPromises).then(() => {
      // collapse void[] to void
    });
  }

  private async _executeTaskAndChainAsync(task: Task): Promise<void> {
    const context: IBuilderContext = {
      stdioSummarizer: task.stdioSummarizer,
      collatedWriter: task.collatedWriter,
      quietMode: this._quietMode
    };

    try {
      const result: TaskStatus = await task.builder.executeAsync(context);

      task.stopwatch.stop();
      task.stdioSummarizer.close();

      this._currentActiveTasks--;
      switch (result) {
        case TaskStatus.Success:
          this._markTaskAsSuccess(task);
          break;
        case TaskStatus.SuccessWithWarning:
          this._hasAnyWarnings = true;
          this._markTaskAsSuccessWithWarning(task);
          break;
        case TaskStatus.Skipped:
          this._markTaskAsSkipped(task);
          break;
        case TaskStatus.Failure:
          this._hasAnyFailures = true;
          this._markTaskAsFailed(task);
          break;
      }
    } catch (error) {
      task.stdioSummarizer.close();

      this._currentActiveTasks--;

      this._hasAnyFailures = true;

      // eslint-disable-next-line require-atomic-updates
      task.error = error;

      this._markTaskAsFailed(task);
    }

    task.collatedWriter.close();

    await this._startAvailableTasksAsync();
  }

  /**
   * Marks a task as having failed and marks each of its dependents as blocked
   */
  private _markTaskAsFailed(task: Task): void {
    task.collatedWriter.terminal.writeStderrLine(colors.red(`"${task.name}" failed to build.`));
    task.status = TaskStatus.Failure;
    task.dependents.forEach((dependent: Task) => {
      this._markTaskAsBlocked(dependent, task);
    });
  }

  /**
   * Marks a task and all its dependents as blocked
   */
  private _markTaskAsBlocked(task: Task, failedTask: Task): void {
    if (task.status === TaskStatus.Ready) {
      this._completedTasks++;
      task.collatedWriter.terminal.writeStderrLine(`"${task.name}" is blocked by "${failedTask.name}".`);
      task.status = TaskStatus.Blocked;
      task.dependents.forEach((dependent: Task) => {
        this._markTaskAsBlocked(dependent, failedTask);
      });
    }
  }

  /**
   * Marks a task as being completed, and removes it from the dependencies list of all its dependents
   */
  private _markTaskAsSuccess(task: Task): void {
    if (task.builder.hadEmptyScript) {
      task.collatedWriter.terminal.writeStdoutLine(colors.green(`"${task.name}" had an empty script.`));
    } else {
      task.collatedWriter.terminal.writeStdoutLine(
        colors.green(`"${task.name}" completed successfully in ${task.stopwatch.toString()}.`)
      );
    }
    task.status = TaskStatus.Success;

    task.dependents.forEach((dependent: Task) => {
      if (!this._changedProjectsOnly) {
        dependent.builder.isIncrementalBuildAllowed = false;
      }
      dependent.dependencies.delete(task);
    });
  }

  /**
   * Marks a task as being completed, but with warnings written to stderr, and removes it from the dependencies
   * list of all its dependents
   */
  private _markTaskAsSuccessWithWarning(task: Task): void {
    task.collatedWriter.terminal.writeStderrLine(
      colors.yellow(`"${task.name}" completed with warnings in ${task.stopwatch.toString()}.`)
    );
    task.status = TaskStatus.SuccessWithWarning;
    task.dependents.forEach((dependent: Task) => {
      if (!this._changedProjectsOnly) {
        dependent.builder.isIncrementalBuildAllowed = false;
      }
      dependent.dependencies.delete(task);
    });
  }

  /**
   * Marks a task as skipped.
   */
  private _markTaskAsSkipped(task: Task): void {
    task.collatedWriter.terminal.writeStdoutLine(colors.green(`${task.name} was skipped.`));
    task.status = TaskStatus.Skipped;
    task.dependents.forEach((dependent: Task) => {
      dependent.dependencies.delete(task);
    });
  }

  /**
   * Prints out a report of the status of each project
   */
  private _printTaskStatus(): void {
    const tasksByStatus: { [status: number]: Task[] } = {};
    this._tasks.forEach((task: Task) => {
      if (tasksByStatus[task.status]) {
        tasksByStatus[task.status].push(task);
      } else {
        tasksByStatus[task.status] = [task];
      }
    });

    this._terminal.writeStderrLine('');

    // These cases should never happen:
    this._printStatus(TaskStatus.Executing, tasksByStatus, colors.red);
    this._printStatus(TaskStatus.Ready, tasksByStatus, colors.red);

    // These are ordered so that the most interesting statuses appear last:
    this._printStatus(TaskStatus.Skipped, tasksByStatus, colors.gray);
    this._printStatus(TaskStatus.Success, tasksByStatus, colors.green);
    this._printStatus(
      TaskStatus.SuccessWithWarning,
      tasksByStatus,
      (text: string) => colors.yellow(text),
      (text: string) => colors.yellow(colors.underline(text))
    );
    this._printStatus(TaskStatus.Blocked, tasksByStatus, colors.red);
    this._printStatus(TaskStatus.Failure, tasksByStatus, colors.red);

    const tasksWithErrors: Task[] = tasksByStatus[TaskStatus.Failure];
    if (tasksWithErrors) {
      tasksWithErrors.forEach((task: Task) => {
        if (task.error) {
          this._terminal.writeStderrLine(`[${task.name}] ${task.error.message}`);
        }
      });
    }

    this._terminal.writeStdoutLine('');
  }

  private _printStatus(
    status: TaskStatus,
    tasksByStatus: { [status: number]: Task[] },
    color: (text: string) => string,
    headingColor: (text: string) => string = color
  ): void {
    const tasks: Task[] = tasksByStatus[status];

    if (tasks && tasks.length) {
      this._terminal.writeStdoutLine(headingColor(`${status} (${tasks.length})`));
      this._terminal.writeStdoutLine(color('================================'));
      for (let i: number = 0; i < tasks.length; i++) {
        const task: Task = tasks[i];

        switch (status) {
          case TaskStatus.Executing:
          case TaskStatus.Ready:
          case TaskStatus.Skipped:
            this._terminal.writeStdoutLine(color(task.name));
            break;

          case TaskStatus.Success:
          case TaskStatus.SuccessWithWarning:
          case TaskStatus.Blocked:
          case TaskStatus.Failure:
            if (task.stopwatch && !task.builder.hadEmptyScript) {
              const time: string = task.stopwatch.toString();
              this._terminal.writeStdoutLine(headingColor(`${task.name} (${time})`));
            } else {
              this._terminal.writeStdoutLine(headingColor(`${task.name}`));
            }
            break;
        }

        if (task.collatedWriter) {
          const shouldPrintDetails: boolean =
            task.status === TaskStatus.Failure || task.status === TaskStatus.SuccessWithWarning;

          const details: string = task.stdioSummarizer.getReport();
          if (details && shouldPrintDetails) {
            // Don't write a newline, because the report will always end with a newline
            this._terminal.writeChunk({ text: details, kind: TerminalChunkKind.Stdout });
          }
        }
      }

      this._terminal.writeStdoutLine(color('================================' + os.EOL));
    }
  }
}
