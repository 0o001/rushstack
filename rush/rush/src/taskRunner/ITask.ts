/**
 * @file ITask.ts
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 *
 * The definition of a task as used by the TaskRunner
 */

import TaskError from '../errorDetection/TaskError';
import TaskStatus from './TaskStatus';
import { ITaskWriter } from './TaskWriterFactory';

/**
 * A definition for a task, an execute function returning a promise and a unique string name
 */
export interface ITaskDefinition {
  name: string;
  execute: (writer: ITaskWriter) => Promise<void>;
}

/**
 * The interface used internally by TaskRunner, which tracks the dependencies and execution status
 */
export interface ITask extends ITaskDefinition {
  status: TaskStatus;
  dependencies: ITask[];
  dependents: ITask[];
  errors: TaskError[];
}
export default ITask;
