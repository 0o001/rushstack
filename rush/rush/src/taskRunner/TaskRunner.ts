/**
 * @file TaskRunner.ts
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 *
 * Executes an arbitrary sequence of tasks based on their dependency graph
 */

import * as colors from 'colors';
import * as os from 'os';

import ITask, { ITaskDefinition } from './ITask';
import TaskStatus from './TaskStatus';
import TaskError from '../errorDetection/TaskError';
import ConsoleModerator, { DualTaskStream } from '@ms/console-moderator';
import { ErrorDetectionMode } from '../errorDetection/ErrorDetector';

/**
 * A class which manages the execution of a set of tasks with interdependencies.
 * Any class of task definition may be registered, and dependencies between tasks are
 * easily specified. Initially, and at the end of each task execution, all unblocked tasks
 * are added to a ready queue which is then executed. This is done continually until all
 * tasks are complete, or prematurely fails if any of the tasks fail. Note that all task
 * definitions must
 * @todo #168352: add unit tests
 */
export default class TaskRunner {
  private _tasks: Map<string, ITask>;
  private _readyTaskQueue: ITask[];
  private _quietMode: boolean;
  private _hasAnyFailures: boolean;
  private _moderator: ConsoleModerator<DualTaskStream>;

  constructor(quietMode: boolean = false) {
    this._tasks = new Map<string, ITask>();
    this._readyTaskQueue = [];
    this._quietMode = quietMode;
    this._hasAnyFailures = false;
    this._moderator = new ConsoleModerator<DualTaskStream>();
  }

  /**
   * Registers a task definition to the map of defined tasks
   */
  public addTask(taskDefinition: ITaskDefinition): void {
    if (this._tasks.has(taskDefinition.name)) {
      throw new Error('A task with that name has already been registered.');
    }

    // @todo #168287: do a copy here
    const task: ITask = taskDefinition as ITask;
    task.dependencies = [];
    task.dependents = [];
    task.errors = [];
    task.status = TaskStatus.Ready;
    this._tasks.set(task.name, task);
  }

  /**
   * Defines the list of dependencies for an individual task.
   * @param taskName - the string name of the task for which we are defining dependencies. A task with this
   * name must already have been registered.
   * @taskDependencies
   */
  public addDependencies(taskName: string, taskDependencies: string[]): void {
    const task: ITask = this._tasks.get(taskName);

    if (!task) {
      throw new Error(`The task '${taskName}' has not been registered`);
    }
    if (!taskDependencies) {
      throw new Error('The list of dependencies must be defined');
    }

    taskDependencies.forEach((dependencyName: string) => {
      const dependency: ITask = this._tasks.get(dependencyName);
      task.dependencies.push(dependency);
      dependency.dependents.push(task);
    });
  }

  /**
   * Executes all tasks which have been registered, returning a promise which is resolved when all the
   * tasks are completed successfully, or rejects when any task fails.
   */
  public execute(): Promise<void> {
    this._tasks.forEach((task: ITask) => {
      if (task.dependencies.length === 0 && task.status === TaskStatus.Ready) {
        this._readyTaskQueue.push(task);
      }
    });

    return new Promise<void>((complete: () => void, reject: () => void) => {
      this._startAvailableTasks(complete, reject);
    });
  }

  /**
   * Helper function which finds any tasks which are available to run and begins executing them.
   * It calls the complete callback when all tasks are completed, or rejects if any task fails.
   */
  private _startAvailableTasks(complete: () => void, reject: () => void): void {
    if (!this._areAnyTasksReadyOrExecuting()) {
      this._printTaskStatus();
      if (this._hasAnyFailures) {
        reject();
      } else {
        complete();
      }
    }

    // @todo #168344: add ability to limit execution to n number of simultaneous tasks
    // @todo #168346: we should sort the ready task queue in such a way that we build projects with deps first
    while (this._readyTaskQueue.length) {
      const task: ITask = this._readyTaskQueue.shift();
      if (task.status === TaskStatus.Ready) {
        task.status = TaskStatus.Executing;
        console.log(colors.yellow(`> Starting task [${task.name}]`));

        const taskStream: DualTaskStream = new DualTaskStream(this._quietMode);

        this._moderator.registerTask(task.name, taskStream);

        task.execute(taskStream)
          .then(() => {
            taskStream.end();
            this._markTaskAsSuccess(task);
            this._startAvailableTasks(complete, reject);
          }).catch((errors: TaskError[]) => {
            taskStream.end();
            this._hasAnyFailures = true;
            task.errors = errors;
            this._markTaskAsFailed(task);
            this._startAvailableTasks(complete, reject);
          });
      }
    }
  }

  /**
   * Marks a task as having failed and marks each of its dependents as blocked
   */
  private _markTaskAsFailed(task: ITask): void {
    console.log(colors.red(`${os.EOL}> Completed task [${task.name}] with errors!`));
    task.status = TaskStatus.Failure;
    task.dependents.forEach((dependent: ITask) => {
      this._markTaskAsBlocked(dependent, task);
    });
  }

  /**
   * Marks a task and all its dependents as blocked
   */
  private _markTaskAsBlocked(task: ITask, failedTask: ITask): void {
    if (task.status === TaskStatus.Ready) {
      console.log(colors.red(`> [${task.name}] blocked by [${failedTask.name}]!`));
      task.status = TaskStatus.Blocked;
      task.dependents.forEach((dependent: ITask) => {
        this._markTaskAsBlocked(dependent, failedTask);
      });
    }
  }

  /**
   * Marks a task as being completed, and removes it from the dependencies list of all its dependents
   */
  private _markTaskAsSuccess(task: ITask): void {
    console.log(colors.green(`> Completed task [${task.name}]`));
    task.status = TaskStatus.Success;
    task.dependents.forEach((dependent: ITask) => {
      const i: number = dependent.dependencies.indexOf(task);
      if (i !== -1) {
        dependent.dependencies.splice(i, 1);
      }

      if (dependent.dependencies.length === 0 && dependent.status === TaskStatus.Ready) {
        this._readyTaskQueue.push(dependent);
      }
    });
  }

  /**
   * Do any Ready or Executing tasks exist?
   */
  private _areAnyTasksReadyOrExecuting(): boolean {
    let anyNonCompletedTasks: boolean = false;
    this._tasks.forEach((task: ITask) => {
      if (task.status === TaskStatus.Executing || task.status === TaskStatus.Ready) {
        anyNonCompletedTasks = true;
      }
    });
    return anyNonCompletedTasks;
  }

  /**
   * Prints out a report of the status of each project
   */
  private _printTaskStatus(): void {
    const tasksByStatus: { [status: number]: ITask[] } = {};
    this._tasks.forEach((task: ITask) => {
      if (tasksByStatus[task.status]) {
        tasksByStatus[task.status].push(task);
      } else {
        tasksByStatus[task.status] = [task];
      }
    });

    console.log('');

    this._printStatus('EXECUTING', tasksByStatus[TaskStatus.Executing], colors.yellow);
    this._printStatus('READY', tasksByStatus[TaskStatus.Ready], colors.white);
    this._printStatus('SUCCESS', tasksByStatus[TaskStatus.Success], colors.green);
    this._printStatus('BLOCKED', tasksByStatus[TaskStatus.Blocked], colors.red);
    this._printStatus('FAILURE', tasksByStatus[TaskStatus.Failure], colors.red);

    const tasksWithErrors: ITask[] = tasksByStatus[TaskStatus.Failure];
    if (tasksWithErrors) {
      tasksWithErrors.forEach((task: ITask) => {
        task.errors.forEach((error: TaskError) => {
          console.log(colors.red(`[${task.name}] ${error.toString(ErrorDetectionMode.LocalBuild) }`));
        });
      });
    }

    console.log('');
  }

  private _printStatus(status: string, tasks: ITask[], color: (a: string) => string): void {
    if (tasks && tasks.length) {
      console.log(color(`${status} (${tasks.length})`));
      console.log(color('================================'));
      for (const task of tasks) {
        console.log(color(task.name));
      }
      console.log(color('================================' + os.EOL));
    }
  }

}
