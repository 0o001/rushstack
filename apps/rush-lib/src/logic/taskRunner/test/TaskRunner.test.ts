import { EOL } from 'os';
import { TaskRunner } from '../TaskRunner';
import { ITaskWriter } from '@microsoft/stream-collator';
import { TaskStatus } from '../TaskStatus';
import { ITaskDefinition } from '../ITask';

class TestConsole {
  public messages: Array<String> = [];

  public log(message: string): void {
    this.messages.push(message);
  }

  public concatenate(): string {
    return this.messages.join(EOL);
  }
}

function createDummyTask(name: string, action?: () => void): ITaskDefinition {
  return {
    name,
    isIncrementalBuildAllowed: false,
    execute: (writer: ITaskWriter) => {
      if (action) {
        action();
      }
      return Promise.resolve(TaskStatus.Success);
    }
  };
}

describe('TaskRunner', () => {
  let logger: TestConsole;
  let taskRunner: TaskRunner;

  describe('Constructor', () => {
    it('throwsErrorOnInvalidParallelism', () => {
      logger = new TestConsole();
      expect(() => new TaskRunner(false, 'tequila', false, logger))
        .toThrowErrorMatchingSnapshot();
    });
  });

  describe('Dependencies', () => {
    beforeEach(() => {
      logger = new TestConsole();
      taskRunner = new TaskRunner(false, '1', false, logger);
    });

    it('throwsErrorOnNonExistentTask', () => {
      expect(() => taskRunner.addDependencies('foo', []))
        .toThrowErrorMatchingSnapshot();
    });

    it('throwsErrorOnNonExistentDependency', () => {
      taskRunner.addTask(createDummyTask('foo'));
      expect(() => taskRunner.addDependencies('foo', ['bar']))
        .toThrowErrorMatchingSnapshot();
    });

    it('detectsDependencyCycle', () => {
      taskRunner.addTask(createDummyTask('foo'));
      taskRunner.addTask(createDummyTask('bar'));
      taskRunner.addDependencies('foo', ['bar']);
      taskRunner.addDependencies('bar', ['foo']);
      expect(() => taskRunner.execute()).toThrowErrorMatchingSnapshot();
    });

    it('respectsDependencyOrder', () => {
      const result: Array<string> = [];
      taskRunner.addTask(createDummyTask('two', () => result.push('2')));
      taskRunner.addTask(createDummyTask('one', () => result.push('1')));
      taskRunner.addDependencies('two', ['one']);
      return taskRunner
        .execute()
        .then(() => {
          expect(result.join(',')).toEqual('1,2');
        })
        .catch(error => fail(error));
    });
  });

  describe('Error logging', () => {
    beforeEach(() => {
      logger = new TestConsole();
      taskRunner = new TaskRunner(false, '1', false, logger);
    });

    const EXPECTED_FAIL: string = 'Promise returned by execute() resolved but was expected to fail';

    it('printedStderrAfterError', () => {
      taskRunner.addTask({
        name: 'stdout+stderr',
        isIncrementalBuildAllowed: false,
        execute: (writer: ITaskWriter) => {
          writer.write('Hold my beer...' + EOL);
          writer.writeError('Woops' + EOL);
          return Promise.resolve(TaskStatus.Failure);
        }
      });
      return taskRunner
        .execute()
        .then(() => fail(EXPECTED_FAIL))
        .catch(err => {
          expect(err.message).toMatchSnapshot();
          const allMessages: string = logger.concatenate();
          expect(allMessages).not.toContain('Hold my beer...');
          expect(allMessages).toContain('Woops');
        });
    });
  });
});
