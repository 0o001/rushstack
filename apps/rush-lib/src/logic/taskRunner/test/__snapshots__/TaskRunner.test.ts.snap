// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskRunner Constructor throwsErrorOnInvalidParallelism 1`] = `"Invalid parallelism value of 'tequila', expected a number or 'max'"`;

exports[`TaskRunner Dependencies detectsDependencyCycle 1`] = `
"A cyclic dependency was encountered:
  foo
  -> bar
  -> foo
Consider using the cyclicDependencyProjects option for rush.json."
`;

exports[`TaskRunner Dependencies respectsDependencyOrder 1`] = `"Registered two[n]Registered one[n]Executing a maximum of 1 simultaneous processes...[-n-][n][x][37m[one] started[x][39m[n][x][32m1 of 2: [one] completed successfully in 0.00 seconds[x][39m[n][x][37m[two] started[x][39m[n][x][32m2 of 2: [two] completed successfully in 0.00 seconds[x][39m[n][n][x][32mSUCCESS (2)[x][39m[n][x][32m================================[x][39m[n][x][32mtwo (0.00 seconds)[x][39m[n][x][32mone (0.00 seconds)[x][39m[n][x][32m================================[x][39m[-n-][x][32m[x][39m[n][n]"`;

exports[`TaskRunner Dependencies respectsDependencyOrder 2`] = `""`;

exports[`TaskRunner Dependencies respectsDependencyOrder 3`] = `""`;

exports[`TaskRunner Dependencies respectsDependencyOrder 4`] = `""`;

exports[`TaskRunner Dependencies throwsErrorOnNonExistentDependency 1`] = `"The project 'bar' has not been registered."`;

exports[`TaskRunner Dependencies throwsErrorOnNonExistentTask 1`] = `"The task 'foo' has not been registered"`;

exports[`TaskRunner Error logging printedStderrAfterError 1`] = `"Project(s) failed to build"`;

exports[`TaskRunner Error logging printedStderrAfterError 2`] = `"Registered stdout+stderr[n]Executing a maximum of 1 simultaneous processes...[-n-][n][x][37m[stdout+stderr] started[x][39m[n][n][x][31mFAILURE (1)[x][39m[n][x][31m================================[x][39m[n][x][31mstdout+stderr (0.00 seconds)[x][39m[n]Woops[n][x][31m================================[x][39m[-n-][x][31m[x][39m[n][n]"`;

exports[`TaskRunner Error logging printedStderrAfterError 3`] = `""`;

exports[`TaskRunner Error logging printedStderrAfterError 4`] = `""`;

exports[`TaskRunner Error logging printedStderrAfterError 5`] = `"[-n-]1 of 1: [stdout+stderr] failed to build![n]"`;
